networking: add Linux-specific TAP mode to network devices

dougvj opened the pull request:
Adds a Linux-specific TAP device networking mode. This allows 86Box's networking devices to be exposed as local TAP devices on the Linux networking stack. In addition, the device may be optionally added to a provided bridge device. If the bridge device does not exist, it is created. 

This allows simple networking of many 86Box machines by specifying the same bridge device. Real ethernet devices on the host can also be added to the bridge to allow for networking with real hardware. 

Note that the 86Box binary must have CAP_NET_ADMIN set in order to use this mode

### TAP Mode configuration interface

![TAP Device Configuration](https://github.com/86Box/86Box/assets/6045454/8f2f2118-0007-4764-9c5c-e8dc0fed6767)

### Demonstration

This shows a Windows 95 and a Windows NT box networked together using TAP mode. They are connected to the same bridge with a wireshark network capture


![WinNT and Win95 Networked with Wireshark Capture](https://github.com/86Box/86Box/assets/6045454/3dc42965-8d24-4f56-bf85-b5ecc2242395)

cold-brewed commented:
Hi, thanks for the PR!

I'll need a little bit of time to review everything.

Have you by chance been able to test compilation on other platforms (win, mac) with these changes? Mostly to make sure everything compiles and the features are properly excluded. If not, the CI runs should at least verify the compilation aspect.

dougvj commented:
I don't have a convenient dev platform for mac, but I can test windows this evening

dougvj commented:
I pushed some minor fixes that I caught with another look through:

1. Accidentally missing QT label for VDE
2. Switch use of strncpy to snprintf
3. Removed accidentally left in debug statement

I also made sure that the build still works on windows (with MSYS2) and noticed that I hadn't guarded the TAP option in QT settings from the windows build properly, so I also fixed that.

cold-brewed commented:
Unfortunately it looks like your fork caught master at a bad time when there were some build failures. 

Could you update your fork and rebase on current master? That way the CI checks can all run again without build failures (which don't appear to be related to your changes).

dougvj commented:
No problem this is done

MamaMelaTodo commented:
Any update?

dougvj commented:
Rebased against master and also added more useful error messages. 

Note that I think something changed since I worked on this, the binary now requires both NET_CAP_ADMIN and NET_CAP_RAW, I have changed the error message to reflect this

dekart811 commented:
Can this be merged, please?

jriwanek commented:
Sorry about the delay, merged, some improvements to follow shortly
